src/init.cpp

```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-                    chainstate->InitCoinsCache(nCoinCacheUsage);
-                    assert(chainstate->CanFlushToDisk());
-
-                    if (!is_coinsview_empty(chainstate)) {
-                        // LoadChainTip initializes the chain based on CoinsTip()'s best block
-                        if (!chainstate->LoadChainTip()) {
-                            strLoadError = _("Error initializing block database");
-                            failed_chainstate_init = true;
-                            break; // out of the per-chainstate loop
-                        }
-                        assert(chainstate->m_chain.Tip() != nullptr);
-                    }
-                }
-
-                if (failed_chainstate_init) {
-                    break; // out of the chainstate activation do-while
-                }
-            } catch (const std::exception& e) {
-                LogPrintf("%s\n", e.what());
-                strLoadError = _("Error opening block database");
+        const int64_t load_block_index_start_time = GetTimeMillis();
+        std::optional<ChainstateLoadingError> rv;
+        try {
+            rv = LoadChainstateSequence(fReset,
+                                        chainman,
+                                        Assert(node.mempool.get()),
+                                        fPruneMode,
+                                        chainparams.GetConsensus(),
+                                        fReindexChainState,
+                                        cache_sizes.block_tree_db_cache_size,
+                                        cache_sizes.coin_db_cache_size,
+                                        cache_sizes.coin_cache_usage_size,
+                                        args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS),
+                                        args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL),
+                                        false,
+                                        false,
+                                        GetAdjustedTime,
+                                        ShutdownRequested,
+                                        []() {
+                                            uiInterface.ThreadSafeMessageBox(
+                                                _("Error reading from database, shutting down."),
+                                                "", CClientUIInterface::MSG_ERROR);
+                                        },
+                                        []() {
+                                            uiInterface.InitMessage(_("Verifying blocks…").translated);
+                                        });
+        } catch (const std::exception& e) {
+            LogPrintf("%s\n", e.what()); // XXX
+            rv = ChainstateLoadingError::ERROR_GENERIC_BLOCKDB_OPEN_FAILED;
+        }
+        if (rv.has_value()) {
+            switch (rv.value()) {
+            case ChainstateLoadingError::ERROR_LOADING_BLOCK_DB:
```

Comment:
I agree with @theuni but in any case this feels like bikeshedding to me. What's written as-is is a clear improvement and refining this further should be optional or follow-up IMO.

Reply:
>  How about a lookup function rather than marrying them? Something like master/src/script/script_error.cpp#L10

I agree that this would be an improvement, separating the error code:string lookup from the recovery logic.

Reply:
In commit "init/chainstate: Decouple from stringy errors" (cd29d334a38fdaf0bbbe810a9cbdeeffad394daa)

re: https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743709991

I agree with John's original comment, and do not think fine grained error codes are good or that some kind of external error code -> string indirection would be good. I think the ideal thing would be for this to return a straightforward SUCCESS / FAILURE / INTERRUPTED status, with a detailed error string or list of error strings in the case of failure. The problem with fine grained return codes is:

- They are maintenance burden making code more verbose than it needs to be and hard to spot inconsistencies easier to introduce.
- They are unergnomic for callers (important if this is supposed evolve into a libbitcoin kernel API) because it's more work for callers to figure out how to handle a long list error codes correctly than a small set of cases.
- They are worse for user experience because it's not possible to include specific information about errors that may be helpful like filenames, timestamps, hashes.